packages/core/src/ctype/CType.chain.ts

```diff
@@ -33,12 +44,108 @@ export function idToChain(cTypeId: ICType['$id']): CTypeHash {
   return idToHash(cTypeId)
 }
 
+// Transform a blockchain-formatted CType input (represented as Bytes) into the original [[IPublicCredentialInput]].
+// It throws if what was written on the chain was garbage.
+function cTypeInputFromChain(input: Bytes): ICType {
+  try {
+    // Throws on invalid JSON input. CType is expected to be a valid JSON document.
+    const reconstructedObject = JSON.parse(input.toUtf8())
+    // Re-compute the ID to validate the resulting ICType.
+    const reconstructedCTypeId = hashToId(getHashForSchema(reconstructedObject))
+    const reconstructedCType: ICType = {
+      ...reconstructedObject,
+      $id: reconstructedCTypeId,
+    }
+    // If throws if the input was a valid JSON but not a valid CType.
+    verifyDataStructure(reconstructedCType)
+    return reconstructedCType
+  } catch (cause) {
+    throw new SDKErrors.CTypeError(
+      `The provided payload cannot be parsed as a CType: ${input.toHuman()}`,
+      { cause }
+    )
+  }
+}
+
+/**
+ * The complete details of a CType, which combines on-chain and off-chain information.
+ */
+export interface ICTypeDetails extends ICType {
+  /**
+   * The DID of the CType's creator.
+   */
+  creator: DidUri
+  /**
+   * The block number in which the CType was created.
+   */
+  createdAt: BN
+}
+
 /**
- * Decodes the owner DID from the return value of `api.query.ctype.ctypes(ctypeHash)`.
+ * Decodes the CType details returned by `api.query.ctype.ctypes()`.
  *
- * @param encoded The data from the blockchain.
- * @returns The owner DID.
+ * @param cTypeId CType ID to use for the query. It is required to complement the information stored on the blockchain in a [[CtypeCtypeEntry]].
+ * @param cTypeEntry The raw CType details from blockchain.
+ * @returns The [[ICTypeDetails]] as the result of combining the on-chain information and the information present in the tx history.
  */
-export function fromChain(encoded: Option<AccountId>): DidUri {
-  return Did.fromChain(encoded.unwrap())
+export async function fromChain(
```

Comment:
I also thought about it, and wanted to do the same for the public credentials, since we do something very similar. In my opinion, we should get rid of the need to call `api.query` or `api.call` for those places that need to do something else with the result anyway, namely ctypes and public credentials. What do you think?

Reply:
As far as they are comparable, yes, that would be my preference. I know very little about how it's done with the public credentials though